velox/docs/develop/types.rst

```diff
@@ -2,54 +2,107 @@
-SMALLINT                int16_t                         2
-INTEGER                 int32_t	                        4
-BIGINT                  int64_t                         8
-DATE                    struct Date                     8
-REAL                    float                           4
-DOUBLE                  double                          8
-SHORT_DECIMAL           struct UnscaledShortDecimal     8
-LONG_DECIMAL            struct UnscaledLongDecimal     16
-TIMESTAMP               struct Timestamp               16
-INTERVAL DAY TO SECOND  struct IntervalDayTime          8
-VARCHAR                 struct StringView              16
-VARBINARY               struct StringView              16
-======================  ===========================    ==================
-
-All scalar types except DECIMAL have a one-to-one mapping with their C++ types.
-
-DECIMAL type is a special scalar type since it carries additional `precision`,
-and `scale` information. Similar to SQL decimal, `precision` is the number of
+Physical Types
+~~~~~~~~~~~~~~
+Each physical type is implemented using a C++ type. The table
+below shows the supported physical types, their corresponding C++ type,
+and bytes required per value.
+
+================   ===========================   ===================
+Physical Type      C++ Type                      Bytes per Value
+================   ===========================   ===================
+BOOLEAN            bool                          0.125 (i.e. 1 bit)
+TINYINT            int8_t                        1
+SMALLINT           int16_t                       2
+INTEGER            int32_t	                     4
+BIGINT             int64_t                       8
+HUGEINT            int128_t                      16
+DATE               struct Date                   8
+REAL               float                         4
+DOUBLE             double                        8
+TIMESTAMP          struct Timestamp              16
+VARCHAR            struct StringView             16
+VARBINARY          struct StringView             16
+OPAQUE             std::shared_ptr<void>         User-defined
+UNKNOWN            struct UnknownValue           0
+================   ===========================   ===================
+
+All physical types have a one-to-one mapping with their C++ types.
+The C++ type is also used as a template parameter for vector classes.
+
+StringView is a struct that contains a 4-byte size field, a 4-byte prefix field,
+and an 8-byte field to store the remaining string (after the prefix) in-line or
+a pointer to the full string out-of-line.
+Refer :doc:`vectors </develop/vectors>` for how vectors of StringView are stored.
+
+Velox allows users to define custom types by extending the OPAQUE type.
+An opaque type must be specified wih an unique `std::type_index`.
+Values for this type must be provided as `std::shared_ptr<T>` where T is a C++ type.
+
+UNKNOWN types are commomly used when type inference fails during runtime.
```

Comment:
That's not correct. UNKNOWN type is used to represent an empty or all nulls vector of unknown type. For example, SELECT array() returns an ARRAY(UNKNOWN()) because it is not possible to determine the type of the elements. This works because there are no elements.

Reply:
> I see that UNKNOWN type is also used during Variant type inference

That's a bug. We should remove this logic altogether since it is not possible to infer logical type from physical type.